modtools/create-unit.lua

```diff
@@ -37,64 +27,94 @@ Creates a unit.  Usage::
         examples:
             DWARF
             HUMAN
+
     -caste casteName
         specify the caste of the unit to be created
-        if not specified, the caste is randomly selected
+        if omitted, the caste is randomly selected
         examples:
             MALE
             FEMALE
             DEFAULT
+
     -domesticate
         tames the unit if it lacks the CAN_LEARN and CAN_SPEAK tokens
+
     -civId id
         Make the created unit a member of the specified civ
         (or none if id = -1).  If id is \\LOCAL, make it a member of the
         civ associated with the fort; otherwise id must be an integer
+
     -groupId id
         Make the created unit a member of the specified group
         (or none if id = -1).  If id is \\LOCAL, make it a member of the
         group associated with the fort; otherwise id must be an integer
+
     -setUnitToFort
         Sets the groupId and civId to the local fort
         Can be used instead of -civId \\LOCAL and -groupId \\LOCAL
+
     -name entityRawName
         Set the unit's name to be a random name appropriate for the
         given entity. \\LOCAL can be specified instead to automatically
         use the fort group entity in fortress mode only
         examples:
             MOUNTAIN
             EVIL
+
     -nick nickname
         set the unit's nickname directly
+
     -location [ x y z ]
         create the unit at the specified coordinates
+
     -age howOld
         set the birth date of the unit by current age
         chosen randomly if not specified
+
     -quantity howMany
         replace "howMany" with the number of creatures you want to create
         defaults to 1 if not specified
+
     -flagSet [ flag1 flag2 ... ]
         set the specified unit flags in the new unit to true
         flags may be selected from df.unit_flags1, df.unit_flags2,
         or df.unit_flags3
+
     -flagClear [ flag1 flag2 ... ]
         set the specified unit flags in the new unit to false
         flags may be selected from df.unit_flags1, df.unit_flags2,
         or df.unit_flags3
 
 ]====]
 
---[[
-if dfhack.gui.getCurViewscreen()._type ~= df.viewscreen_dwarfmodest or df.global.ui.main.mode ~= df.ui_sidebar_mode.LookAround then
-  print 'activate loo[k] mode'
-  return
-end
---]]
+local utils = require 'utils'
 
-local utils=require 'utils'
+function createUnit(raceStr, casteStr, pos, age, domesticate, civ_id, group_id, entityRawName, nickname, quantity, flagSet, flagClear)
+--  creates the desired unit(s) at the specified location
+--  returns a table containing the created unit(s)
+  if not pos then
+    qerror("Location not specified!") -- check repeated for module usage
```

Comment:
This is breaking `spawnunit`. Haven't investigated further, but maybe `pos` used to default to the cursor position?

Reply:
The createUnitInner function in modtools/create-unit works by simulating the arena mode spawning process roughly as follows: 
1) Populate the arena spawn data as desired
2) Set up a dwarf mode viewscreen
3) Set the ui mode to "look around", causing the cursor to appear (at its last position)
4) Switch to arena mode
5) Move the cursor to the target spawn location
6) Open the arena creature creation menu, with the target creature pre-selected
7) Pass the equivalent of the "Enter" key to spawn the unit
8) Process the unit as desired
9) Loop steps 6-8 if creating multiple units simultaneously
10) Return everything to its original state

As you can see, the position at which the unit is spawned depends entirely on the cursor location prior to opening the creature creation menu. If the target location were to be left unspecified, the cursor would be left untouched, causing the spawn location to default to wherever it was that the cursor was left prior to running the script.

Since such a scenario is generally not desirable, I added a pos check within the function itself just in case it’s ever called by other scripts. `spawnunit` is failing at this check as it does not require the user to pass location data to `modtools/create-unit`, instead seeming to rely on users having selected the spawn location via the cursor prior to running the script  in the event that this information is omitted (despite this not being explained in the documentation). On that note, I have not figured out how to correctly specify a target location using x, y, z coordinates in `spawnunit`, despite this purportedly being possible according to the documentation.

I’m not sure what `spawnunit` is meant to be used for (indeed, I was unaware of its existence prior to reading the above), but given its limited capabilities as a front-end, my initial impulse is to just have it include a cursor check and do away with its defective(?) location input (since end-users are unlikely to fumble around with coordinates in lieu of using the cursor, and modders are better suited to using the proper modtools script for their purposes). Thoughts?

Reply:
The original spawnunit took race, caste, and name, and created a unit at the selected cursor position. The new one is just a wrapper around this script which passes through any extra arguments. If always passing a position is necessary, that shouldn't be too hard, I guess.